clippy_lints/src/redundant_local.rs

```diff
@@ -0,0 +1,136 @@
+
+    fn check_fn(&mut self, cx: &EarlyContext<'_>, func: FnKind<'_>, _span: Span, _node: rustc_ast::NodeId) {
+        // get the top-level block of the function's expression
+        let block = match func {
+            FnKind::Fn(_, _, _, _, _, Some(block)) => block,
+            FnKind::Closure(
+                _,
+                _,
+                Expr {
+                    kind: ExprKind::Block(block, _),
+                    ..
+                },
+            ) => block,
+            _ => return,
+        };
+
+        // collect the initial bindings from the function's parameters
+        let mut bindings = FxHashMap::default();
+        for param in &func.decl().inputs {
+            collect_introduced_bindings(&mut bindings, &param.pat);
+        }
+
+        check_block_with_bindings(cx, &mut bindings, block);
+    }
+}
+
+#[derive(Clone, Copy)]
+struct BindingInfo {
+    mutability: Mutability,
+    definition_span: Span,
+}
+
+// check the block for redundant locals, starting with the given bindings
+fn check_block_with_bindings(cx: &EarlyContext<'_>, bindings: &mut FxHashMap<Ident, BindingInfo>, block: &Block) {
+    if block.span.from_expansion() {
+        return;
+    }
+
+    for stmt in &block.stmts {
+        if let StmtKind::Local(local) = &stmt.kind {
+            if_chain! {
+                if let PatKind::Ident(BindingAnnotation(_, mutability), ident, _) = local.pat.kind;
+                if let LocalKind::Init(expr) = &local.kind;
+                if let ExprKind::Path(None, Path { segments, .. }) = &expr.kind;
+                if segments.len() == 1 && segments[0].ident == ident;
+                if let Some(prev) = bindings.get(&ident);
+                if mutability == prev.mutability;
+                if is_local_user_controlled(cx, local);
+                then {
+                    span_lint_and_help(
+                        cx,
+                        REDUNDANT_LOCAL,
+                        vec![prev.definition_span, local.span],
+                        "redundant redefinition of a binding",
+                        None,
+                        &format!("remove the redefinition of `{ident}`"),
+                    );
+                }
+            }
+
+            collect_introduced_bindings(bindings, &local.pat);
+        }
+    }
+}
+
+// collect the bindings introduced by the given pattern into the passed set of bindings
+fn collect_introduced_bindings(bindings: &mut FxHashMap<Ident, BindingInfo>, pat: &Pat) {
+    pat.walk(&mut |subpat| {
+        if let PatKind::Ident(BindingAnnotation(_, mutability), ident, _) = subpat.kind {
+            bindings.insert(
+                ident,
+                BindingInfo {
+                    mutability,
+                    definition_span: subpat.span,
+                },
+            );
+        }
+        true
+    });
+}
+
+// check if a local is actually controlled by the user
+// otherwise, (macro or compiler generated), we should not lint it
```

Comment:
nit: this should be a doc comment and "compiler generated" can be removed afaik, as compiler generated locals are not visited (or even available in the AST)

Reply:
you should also do this for the other functions here, `collect_introduced_bindings` and `check_block_with_bindings`